### PR TITLE
Remove io::util::Async{Read,Write}Ext

### PR DESCRIPTION
**Pull request by aknuds1**
## Motivation

Remove io::util::Async{Read,Write}Ext traits as stipulated in #3005.

## Solution

I'm moving methods in io::util::Async{Read,Write}Ext traits to corresponding io::Async{Read,Write} traits.

### CODE REVIEW

File: tokio-util/tests/framed_write.rs [hunk 1]
```diff
@@ -155,13 +155,13 @@ impl AsyncWrite for Mock {
         _cx: &mut Context<'_>,
         buf: &[u8],
     ) -> Poll<Result<usize, io::Error>> {
-        match Pin::get_mut(self).write(buf) {
+        match std::io::Write::write(&mut Pin::get_mut(self), buf) {
```
Comment (LucioFranco) [inline tokio-util/tests/framed_write.rs:158]: import `use std::io::Write` above this line?

File: tokio/src/io/async_read.rs [hunk 2]
```diff
@@ -55,6 +85,1057 @@ pub trait AsyncRead {
...
+    ///     // read up to 10 bytes
+    ///     let n = f.read(&mut buffer[..]).await?;
+    ///
+    ///     println!("The bytes: {:?}", &buffer[..n]);
+    ///     Ok(())
+    /// }
+    /// ```
+    fn read<'a>(&'a mut self, buf: &'a mut [u8]) -> Read<'a, Self>
+    where
+        Self: Unpin + Sized,
```
Comment (LucioFranco) [inline tokio/src/io/async_read.rs:188]:
```suggestion
        Self: Sized + Unpin,
```

small nit, I prefer this order

File: tokio/src/io/async_read.rs [hunk 3]
```diff
@@ -55,6 +85,1057 @@ pub trait AsyncRead {
...
+    ///     // read up to 10 bytes, note that the return value is not needed
+    ///     // to access the data that was read as `buffer`'s internal
+    ///     // cursor is updated.
+    ///     f.read_buf(&mut buffer).await?;
+    ///
+    ///     println!("The bytes: {:?}", &buffer[..]);
+    ///     Ok(())
+    /// }
+    /// ```
+    fn read_buf<'a, B>(&'a mut self, buf: &'a mut B) -> async_read_futures::ReadBuf<'a, Self, B>
```
Comment (LucioFranco) [inline tokio/src/io/async_read.rs:250]: @Darksonn @carllerche feel like we should rename this future?

File: tokio/src/io/async_read.rs [hunk 4]
```diff
@@ -55,6 +85,1057 @@ pub trait AsyncRead {
...
+    /// }
+    /// ```
+    ///
+    /// (See also the [`tokio::fs::read`] convenience function for reading from a
+    /// file.)
+    ///
+    /// [`tokio::fs::read`]: fn@crate::fs::read
+    fn read_to_end<'a>(&'a mut self, buf: &'a mut Vec<u8>) -> ReadToEnd<'a, Self>
+    where
+        Self: Unpin + Sized,
```
Comment (LucioFranco) [inline tokio/src/io/async_read.rs:1050]:
```suggestion
        Self: Sized + Unpin,
```

File: tokio/src/io/async_read.rs [hunk 5]
```diff
@@ -55,6 +85,1057 @@ pub trait AsyncRead {
...
+    /// }
+    /// ```
+    ///
+    /// (See also the [`crate::fs::read_to_string`] convenience function for
+    /// reading from a file.)
+    ///
+    /// [`crate::fs::read_to_string`]: fn@crate::fs::read_to_string
+    fn read_to_string<'a>(&'a mut self, dst: &'a mut String) -> ReadToString<'a, Self>
+    where
+        Self: Unpin + Sized,
```
Comment (LucioFranco) [inline tokio/src/io/async_read.rs:1097]:
```suggestion
        Self: Sized + Unpin,
```

File: tokio/src/io/async_read_futures.rs [hunk 6]
```diff
@@ -1,4 +1,4 @@
-use crate::io::AsyncRead;
+use super::AsyncRead;
```
Comment (LucioFranco) [inline tokio/src/io/async_read_futures.rs:1]: Does this file contain other `Read` futures in it?

File: tokio/src/io/async_write.rs [hunk 7]
```diff
@@ -176,6 +209,941 @@ pub trait AsyncWrite {
...
+    ///     let mut file = File::create("foo.txt").await?;
+    ///
+    ///     // Writes some prefix of the byte string, not necessarily all of it.
+    ///     file.write(b"some bytes").await?;
+    ///     Ok(())
+    /// }
+    /// ```
+    fn write<'a>(&'a mut self, src: &'a [u8]) -> Write<'a, Self>
+    where
+        Self: Unpin + Sized,
```
Comment (LucioFranco) [inline tokio/src/io/async_write.rs:261]:
```suggestion
        Self: Sized + Unpin,
```

File: tokio/src/io/async_write.rs [hunk 8]
```diff
@@ -176,6 +209,941 @@ pub trait AsyncWrite {
...
+    ///
+    ///     buffer.write_all(b"some bytes").await?;
+    ///     Ok(())
+    /// }
+    /// ```
+    ///
+    /// [`write`]: AsyncWrite::write
+    fn write_all<'a>(&'a mut self, src: &'a [u8]) -> WriteAll<'a, Self>
+    where
+        Self: Unpin + Sized,
```
Comment (LucioFranco) [inline tokio/src/io/async_write.rs:374]:
```suggestion
        Self: Sized + Unpin,
```

File: tokio/src/io/async_write.rs [hunk 9]
```diff
@@ -176,6 +209,941 @@ pub trait AsyncWrite {
...
+    ///     let mut buffer = BufWriter::new(f);
+    ///
+    ///     buffer.write_all(b"some bytes").await?;
+    ///     buffer.flush().await?;
+    ///     Ok(())
+    /// }
+    /// ```
+    fn flush(&mut self) -> Flush<'_, Self>
+    where
+        Self: Unpin + Sized,
```
Comment (LucioFranco) [inline tokio/src/io/async_write.rs:1104]:
```suggestion
        Self: Sized + Unpin,
```

File: tokio/src/io/async_write.rs [hunk 10]
```diff
@@ -176,6 +209,941 @@ pub trait AsyncWrite {
...
+    ///     let mut buffer = BufWriter::new(f);
+    ///
+    ///     buffer.write_all(b"some bytes").await?;
+    ///     buffer.shutdown().await?;
+    ///     Ok(())
+    /// }
+    /// ```
+    fn shutdown(&mut self) -> Shutdown<'_, Self>
+    where
+        Self: Unpin + Sized,
```
Comment (LucioFranco) [inline tokio/src/io/async_write.rs:1143]:
```suggestion
        Self: Sized + Unpin,
```

**Review by LucioFranco — COMMENTED** [review]
Looking good so far! Couple nits, I'd like @Darksonn and @carllerche to review this as well.

File: tokio-util/tests/framed_write.rs (hunk 1 above)
Comment (aknuds1) [reply tokio-util/tests/framed_write.rs:158]: @LucioFranco OK, will try.

File: tokio/src/io/async_read_futures.rs (hunk 6 above)
Comment (aknuds1) [reply tokio/src/io/async_read_futures.rs:1]: @LucioFranco Just the one, I named it in plural to allow for more. Want me to rename it to async_read_future.rs or something?

File: tokio-util/tests/framed_write.rs (hunk 1 above)
Comment (aknuds1) [reply tokio-util/tests/framed_write.rs:158]: I could actually remove the `std::io::` qualifier, it's no longer ambiguous for some reason (used to be).

File: tokio/src/io/async_read.rs (hunk 3 above)
Comment (aknuds1) [reply tokio/src/io/async_read.rs:250]: That would probably make sense @LucioFranco, but I'll leave it to the team to figure out what the name ought to be.

Comment (carllerche) [reply tokio/src/io/async_read.rs:250]: The future is not public as of now. I would not worry about a rename yet. This is fine.

File: tokio/src/io/async_read_futures.rs (hunk 6 above)
Comment (carllerche) [reply tokio/src/io/async_read_futures.rs:1]: It's internal, not a big deal. We can update in a follow up PR.

My preference would be to create `tokio::io::futures` as a dir and have all the futures in there.

**Review by carllerche — COMMENTED** [review]
Looks good to me :+1: We just need CI to pass. Please avoid any rebasing when pushing to this branch so we can incrementally review changes.

**Comment by aknuds1** [comment]
> Looks good to me  We just need CI to pass. Please avoid any rebasing when pushing to this branch so we can incrementally review changes.

Thanks @carllerche - is there anything I should do?

File: tokio/src/io/mod.rs [hunk 11]
```diff
@@ -17,8 +17,8 @@
 //!
 //! Another difference is that `AsyncRead` and `AsyncWrite` only contain
 //! core methods needed to provide asynchronous reading and writing
-//! functionality. Instead, utility methods are defined in the [`AsyncReadExt`]
-//! and [`AsyncWriteExt`] extension traits. These traits are automatically
+//! functionality. Instead, utility methods are defined in the [`AsyncRead`]
+//! and [`AsyncWrite`] extension traits. These traits are automatically
 //! implemented for all values that implement `AsyncRead` and `AsyncWrite`
 //! respectively.
```
Comment (Darksonn) [inline tokio/src/io/mod.rs:23]: Just delete this section.

File: tokio/src/io/mod.rs [hunk 12]
```diff
@@ -199,6 +198,26 @@ pub use self::async_write::AsyncWrite;
...
+mod shutdown;
+mod write;
+mod write_all;
+mod write_buf;
+mod write_int;
+// Used by process
+pub(crate) use read_to_end::read_to_end;
+mod read_to_string;
+mod take;
+pub use take::Take;
```
Comment (Darksonn) [inline tokio/src/io/mod.rs:216]: This one is public but nothing else is?

File: tokio/tests/uds_split.rs [hunk 13]
```diff
@@ -30,8 +30,8 @@ async fn split() -> std::io::Result<()> {
 }
 
 async fn send_recv_all(
-    read: &mut (dyn AsyncRead + Unpin),
-    write: &mut (dyn AsyncWrite + Unpin),
+    read: &mut (impl AsyncRead + Unpin),
+    write: &mut (impl AsyncWrite + Unpin),
```
Comment (Darksonn) [inline tokio/tests/uds_split.rs:34]: I assume this change is from when you had troubles with object safety. It should not be necessary.

**Review by Darksonn — COMMENTED** [review]
This is my first pass that looks at files other than `async_read.rs` and `async_write.rs`.

**Comment by Darksonn** [comment]
There are quite a lot of errors in the version of CI that checks each feature separately.

1. To conditionally import something depending on features, you can do `#[cfg(feature = "io-util")]` on the use statement. You can also do `#[cfg(all(feature = "io-util", feature = "net"))]` for multiple requirements, or use `any` for an or clause.
2. To conditionally include a method on a trait, use the same `#[cfg(...)]` annotation as above. Additionally you can add an `#[doc(cfg(the_same_thing))]` to add the blue annotation in the docs regarding the required features.

To compile the code locally with only some features enabled, do this:
```
cargo build --features "io-util net"
```
Let me know if you need any help with resolving any of the errors reported by CI.

File: tokio/tests/uds_split.rs (hunk 13 above)
Comment (carllerche) [reply tokio/tests/uds_split.rs:34]: I tried reverting:

```rust
error: the `write_all` method cannot be invoked on a trait object
   --> tokio/tests/uds_split.rs:37:11
    |
37  |     write.write_all(input).await?;
    |           ^^^^^^^^^
    | 
   ::: /home/carllerche/Code/tokio/tokio/src/io/async_write.rs:379:15
    |
379 |         Self: Sized + Unpin,
    |               ----- this has a `Sized` requirement
```

Removing the `Sized` requirement on `fn write_all` prevents `AsyncRead` from becoming a trait object itself.

Comment (carllerche) [reply tokio/tests/uds_split.rs:34]: Same problem as here: https://docs.rs/async-trait/0.1.36/async_trait/#dyn-traits

**Comment by carllerche** [comment]
It seems that getting rid of `AsyncReadExt` is not as simple as we initially expected. There are some object safety issues (see inline comments).

We probably are going to have to abandon this approach and just rename `AsyncReadExt` to -> `AsyncReadUtil` or just `tokio::io::ReadUtil`. Thoughts @tokio-rs/maintainers ?

**Comment by taiki-e** [comment]
> We probably are going to have to abandon this approach and just rename `AsyncReadExt` to -> `AsyncReadUtil` or just `tokio::io::ReadUtil`.

FYI: `Ext` is the suffix recommended by [RFC 445 "Extension Trait Conventions"](https://github.com/rust-lang/rfcs/blob/master/text/0445-extension-trait-conventions.md).

> ## The convention
> ... 
> For true extension traits, there should be a clear type or trait that they are extending. **The extension trait should be called FooExt where Foo is that type or trait.**
> 
> In some cases, the extension trait only applies conditionally. For example, AdditiveIterator is an extension trait currently in std that applies to iterators over numeric types. These extension traits should follow a similar convention, putting together the type/trait name and the qualifications, **together with the Ext suffix**: IteratorAddExt.


> ## Alternatives
> It seems clear that we need some convention here. **Other possible suffixes would be Util or Methods, but Ext is both shorter and connects to the name of the pattern.**

**Comment by carllerche** [comment]
Welp, I'm guessing the answer here is we do nothing...

File: tokio/tests/uds_split.rs (hunk 13 above)
Comment (Darksonn) [reply tokio/tests/uds_split.rs:34]: This compiles, but is not particularly convenient.
```Rust
async fn send_recv_all(
    mut read: &mut (dyn AsyncRead + Unpin),
    mut write: &mut (dyn AsyncWrite + Unpin),
    input: &[u8],
) -> std::io::Result<Vec<u8>> {
    (&mut write).write_all(input).await?;
    (&mut write).shutdown().await?;

    let mut output = Vec::new();
    (&mut read).read_to_end(&mut output).await?;
    Ok(output)
}
```

**Comment by aknuds1** [comment]
> Welp, I'm guessing the answer here is we do nothing...

Already commented in #3005, but I guess we'll have to close the PR? :/

**Comment by carllerche** [comment]
Because of the object safety issues, we are going to keep things as they are. 

@aknuds1 thanks for the work! Even if it didn’t land, it was very helpful to get us to the conclusion.

**Pull request closed without merging**